### ISSUE TITLE
Add a setting to prevent reading observable values outside of actions/reactions

### ISSUE THREAD

**Issue by rrousselGit** [mobxjs/mobx.dart#199]
As discussed on gitter.
Here's the conversation:

>Is there any mean using mobx to warn users of an `@observable` is used outside of a >Reaction/Observer?
>
>It can be quite difficult to see that we forgot watching the store.

>Pavan Podila @pavanpodila 14:12
>Not the reads but writes are protected
>Reads do not affect the consistency of the reactive system

>Remi Rousselet @rrousselGit 14:15
>It's more about catching bugs.
>
>There are many scenarios in Flutter where we can loose reactivity due to a mistake
>
>Something as simple as:
>
>```dart
>Store myStore;
>
>Widget build(BuildContext context) {
>  return AnotherWidget(foo: myStore);
>}
>```
>
> may be bugged, because AnotherWidget may not expect foo to be reactive and therefore not use an Observer.
> 
> Similarly builders are misleading:
> 
> ```dart
> Store myStore;
> 
> Widget build(BuildContext context) => Observer(builder: (_) {
>  return ListView.builder(
>    builder: (context) => Text(myStore.text),
>  ); 
>});
>```
>
>That looks fine, but the Text do not watch changes on myStore.text.

> Michael @smiLLe 14:28
@rrousselGit "may be bugged, because AnotherWidget may not expect foo to be reactive and therefore not use an Observer.” But this has always been a problem with mobx, for example in react

> Remi Rousselet @rrousselGit 14:32
I'm not saying the opposite. I'm just asking if there's a potential solution. 
A setting to prevent reading `@observable` outside of `@action`/Reaction/Observer could solve the issue.

> Michael @smiLLe 14:32
i like the idea

> Pavan Podila @pavanpodila 17:51
That is definitely an interesting thought and worth exploring. I think this is definitely doable, but needs to go behind a setting for sure
Btw we do this already for writes

**Comment by pavanpodila** [comment]
I want to open this up for some discussion, since there is a potential for an API change.

Currently we have a layer of protection for writing to an observable outside of an Action. This is guarded by `enforceActions` on the `ReactiveConfig`. Similarly, we are trying to introduce a behavior where the reads will be protected, or allowed only inside a reaction. I see this as a symmetric behavior.

We can introduce a setting on `ReactiveConfig` to only have reactive-reads. May be we can call it `enforceReadInReaction` with a boolean value. Since we are being explicit about the reads, the original setting of `enforceActions` now appears mis-named. I think we should call it `enforceWriteInActions`

Together these settings will appear symmetric and named appropriately.

```dart
config.enforceReadInAction

config.enforceWriteInReaction


```

What do you guys think?

**Comment by rrousselGit** [comment]
That looks good to me. And we can keep `enforceActions` temporarily as `@deprecated`, to help peoples migrate.

**Comment by pavanpodila** [comment]
Cool...I'll start with a PR on this

**Comment by pavanpodila** [comment]
This has landed with #205
